### ISSUE TITLE
Retro Achievements TODO

### ISSUE THREAD

**Issue by leiradel** [libretro/RetroArch#4867]
* Work offline
  * Cache JSONs with achievements
    * Use game hash as key to avoid needing network to get game ID
  * Cache badges' images
    * Game hash + badge ID
  * Load from cache, fire tasks to update
    * Hot swap in-game if updated?
* Show badges
  * Keep badges in memory during play to avoid hiccups?
    * Atlas?
  * Pre-render text to avoid hiccups?
  * Enter and exit effects
  * How to cope with different resolutions?
* Leaderboards
  * Implement
* Tooling
  * Build tools that use RArch's embedded HTTP server to find achievements
  * Implement the memory layout interface in cores
    * Maybe wait for zeromus' new memory interface? Needs support from Retro Achievements
* New features
  * Rumble (proposed  by @fr500 )
  * Challenges (proposed by @fr500 )

**Comment by meleu** [comment]
Is it fair to request to add "Remove save/load state options from Quick Menu when hardcore mode is enabled" to this list?

**Comment by meleu** [comment]
Hey @leiradel. That leaderboards item can be checked. ;-)

**Comment by leiradel** [comment]
> Is it fair to request to add "Remove save/load state options from Quick Menu when hardcore mode is enabled" to this list?

I thought load/save were already disabled when hardcode mode is on. I'll take a look.

**Comment by meleu** [comment]
Not sure if I was clear. I mean the option appearing on menu when hardcore mode is on.

Currently the option appears, but don't worry, it doesn't work when hardcore mode is on.

The request is just a cosmetic. To be honest I'm not even sure if it worths the work to hide it under that condition.

**Comment by leiradel** [comment]
But the user can select those options? If yes, it's confusing because the user will think that the state is saved and it's not.

So I believe the functions should be there but disabled, and they should include a subtext stating that they're disabled because hardcode mode is on.

**Comment by retro-wertz** [comment]
im confused about leaderboards... should there suppose to be a new entry
for those or even at least a popup? ddnt notice any for nes and snes at
least.

thanks and great work.

**Comment by fr500** [comment]
So far only submitting leaderboards is implemented.
It will give you an osd message when a leaderboard attempt is being
tracked.

Try time trials in super Mario Kart

**Comment by kikeminchas** [comment]
Regarding the offline achievements, I have heard/read conflicting info about the feasibility of this feature. Lots of concern about possible cheating, etc... I personally think it would be a great feature since more and more people is playing on mobile/handheld devices. Is there really a way to implement this on a secure way (or at all, actually). Thanks for the work!

**Comment by leiradel** [comment]
@kikeminchas it's definitely feasible, and some simple anti-cheat measures can be taken. They won't stop all the cheaters, but will prevent the easiest ones.

**Comment by ahstro** [comment]
Just checking: Has offline been implemented? I'd love to be able to play on my Switch on the go 😅

**Comment by mtoensing** [comment]
+1 for offline Achievements on the Switch

**Comment by XargonWan** [comment]
Sorry to post here one year after the last post, but is there any news about the offline achievements?
A lot of retro handled devices doesn't have a constant access to the internet and I think that this would be a great feature.

**Comment by Littlemac123** [comment]
currently how it works is it downloads all the cheevos in RAM and doesn't save that locally for unknown reasons, then as you get the cheevo it connects to 
retroachievements.org/dorequest.php?r=awardachievement&u=`yourname`&t=`yourtoken`&a=`cheevonumber`&h=1
you can lit open this in a browser and unlock anything you want
retroarch could just save this link and go to it when your online again
like honestly idk why this has taken so long to get offline cheevos
it's so simple

**Comment by Jamiras** [comment]
It's not nearly as simple as you think it is. Because of the competitive nature of the site (ranks based on scores), we have to be very careful about any functionality we add that makes it easier for users to game the system.

Offline support leaves a lot of loopholes for cheating. If you decide to store the fact that achievements 3, 4 and 5 have been unlocked, what's to prevent the user from modifying that list to include 6, 7, 8, and 9 too? Or even some other achievement for some game that they're never played?

Another issue is the upfront loading. Many achievement sets are in flux as bugs are found and fixed, or achievements are added or removed. By fetching the most recent version on each login, we ensure the user is playing with the correct version of the achievements. Similarly, validating the game's hash ensures the user isn't playing with a patched ROM where they have infinite lives. Once we start storing the information locally, it becomes trivial for the user to modify it. Sure, we can add checksums or other logic, but with an open source project it's really hard to keep your secrets secret.

As you've identified, it's easy to unlock whatever you want whenever you want, so why should we even have achievements? Let users just rack up points by abusing their browser without even opening the games. But I guess they can't do that offline either, so it doesn't address this ticket. Please remove the URL from your comment to limit the abuse of that loophole.

I would not be against a _completely_ offline implementation, where the achievement definitions were stored and ran offline, but it would not be wise to allow such a configuration to participate in the global rankings for reasons mentioned above. It's bad enough that we have to allow older versions of RetroArch to connect, as they may not contain anti-cheat measures or bugfixes that have been added to newer versions.

For this to truly work, we'd need to capture not only the achievement IDs that were unlocked, but their definitions, and the game hash used while offline and validate the achievement definition and game hash when doing the unlock after going back online. And all of that would have to be done in such a way that the user couldn't easily modify the unlocks to add additional items or spoof the definition/game hash. If you still think that's simple, feel free to open a PR.

**Comment by mtoensing** [comment]
I removed my +1 for this feature. After I read the comment of @Jamiras I think we should leave to as it is. I can always use tethering to my phone to keep my handhelds online. It is better if the system is more secure than to have the comfort of offline usage. At least that is my opinion. I like the competitive nature of the retroachievements. if everyone can unlock every achievement then the whole thing no more fun. Right now wasting 1h hour of my precious time to beat a stage in Super Mario World without a balloon feels like an accomplishment. ;-)

**Comment by Littlemac123** [comment]
>Offline support leaves a lot of loopholes for cheating.
@Jamiras 
good thing it all ready has this regardless if you wish to implement this or not, I been doing cheevos offline then using the URL method when im online just fine to say i got them legit this hole time. idk what your trying to prevent at this point when its all ready there right infront of you

**Comment by Littlemac123** [comment]
im not saying implementing better severity is a bad thing, but currently if your never going to do that regaless or if no ones working on it then what the hell is the harm rn??

**Comment by Jamiras** [comment]
Because right now, not everyone knows how easy it is to generate your own URLs, and I'd like to ask you again to remove that from your earlier comment.

As for future plans, it's hard to say what we're going to do or when we're going to do it. Offline support is pretty low on the list, and there's only a handful of people doing any sort of coding.

**Comment by Littlemac123** [comment]
im sorry but when your highest security is a http get command that even just posting the URL in discord then Discord servers can auto give you a cheevo this is a valid issue that belongs here regardless on its own regardless of the offline implementation I so dream for.

**Comment by Littlemac123** [comment]
if you guys really love valid real users getting real cheevos you should put more sole in your hard work at this point, you been using this for YEARS. it has to stop

**Comment by Jamiras** [comment]
I agree. It would be great to plug that hole, but even if we fixed it today, we couldn't actually enforce it because we allow older versions of RetroArch to continue to talk to the server because many standalone systems (like RetroPie) come prepackaged with a specific version of RetroArch, and their users don't know how to upgrade, or are unwilling to because "if it ain't broke, don't fix it".

And if you don't want your account to become Untracked, I'd recommend you stop circumventing the implementation. We have right in [the documentation](http://docs.retroachievements.org/Global-Leaderboard-and-Achievement-Hunting-Rules/#not-allowed-1) that what you're doing will get you in trouble. You have admitted to violating rule 1, and you're clearly violating rule 9 by not taking down that link.

**Comment by Littlemac123** [comment]
actually I never used it I was just saying that so you can understand a senerio where offline cheevos work, you can see that if you look at my account logs, that I never used older retroarchs other then current, and I'm offended that you removed my post as it's a real issue in your server. as it didn't brake any retroarch rules in this gethub, just your site rules.

**Comment by Littlemac123** [comment]
if you don't like the fact the issue exists then just pull the pug and force current retroarch like lit what's the harm? preventing cheating is good right???

**Comment by Littlemac123** [comment]
ok fine

On April 2, 2021 5:50:39 PM GMT-05:00, Jamiras ***@***.***> wrote:
>I agree. It would be great to plug that hole, but even if we fixed it
>today, we couldn't actually enforce it because we allow older versions
>of RetroArch to continue to talk to the server because many standalone
>systems (like RetroPie) come prepackaged with a specific version of
>RetroArch, and their users don't know how to upgrade, or are unwilling
>to because "if it ain't broke, don't fix it".
>
>And if you don't want your account to become Untracked, I'd recommend
>you stop circumventing the implementation. We have right in [the
>documentation](http://docs.retroachievements.org/Global-Leaderboard-and-Achievement-Hunting-Rules/#not-allowed-1)
>that what you're doing will get you in trouble. You have admitted to
>violating rule 1, and you're clearly violating rule 9 by not taking
>down that link.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/libretro/RetroArch/issues/4867#issuecomment-812745288

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
